### PR TITLE
Anti entropy

### PR DESCRIPTION
**Pull request by Yoga07**

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -26,15 +26,57 @@ pub enum Error {
...
+    #[error("Node does not currently manage any immutable chunks")]
+    NoImmutableChunks,
+    /// Node is currently churning so cannot perform the request.
+    #[error("Cannot complete request due to churning of funds")]
+    NodeChurningFunds,
+    /// Node is currently churning, but failed to sign a message.
+    #[error("Error signing message during churn")]
+    ChurnSignError,
+    /// The node is not in genesis stage.
+    #[error("Not in genesis stage")]
```
Comment (maqi) [inline src/error.rs:60]: Genesis node not in genesis stage.

File: src/error.rs [hunk 2]
```diff
@@ -26,15 +26,57 @@ pub enum Error {
...
     #[error("Not section public key returned from routing")]
     NoSectionPublicKey,
     /// Unknown as a Section PublicKey.
     #[error("PublicKey provided was not identified as a section {0}")]
     UnknownSectionKey(PublicKey),
+    /// Nodes cannot send direct messages
+    #[error("Node cannot send direct messages. This functionality will be deprecated in routing.")]
+    CannotDirectMessage,
+    /// Node cannot be updated, message cannot be resent
+    #[error("Process error could not be handled. We cannot update the problem node.")]
```
Comment (maqi) [inline src/error.rs:78]: the node in problem

File: src/node/handle.rs [hunk 3]
```diff
@@ -7,26 +7,31 @@
...
 use xor_name::XorName;
 
 impl Node {
     ///
-    pub async fn handle(&mut self, duty: NodeDuty) -> Result<NodeDuties> {
+    pub async fn handle(
+        &mut self,
+        duty: NodeDuty,
+        _ctx: &Option<MsgContext>,
+    ) -> Result<NodeDuties> {
```
Comment (maqi) [inline src/node/handle.rs:34]: not used then why pass in?

Comment (maqi) [inline src/node/handle.rs:34]: not used then why pass in?

File: src/node/interaction.rs [hunk 4]
```diff
@@ -58,13 +58,13 @@ impl Node {
...
         Ok(NodeDuty::Send(OutgoingMsg {
-            msg: Message::NodeQueryResponse {
+            msg: MsgType::Node(NodeMsg::NodeQueryResponse {
                 response: NodeQueryResponse::System(NodeSystemQueryResponse::GetSectionElders(
                     elders,
                 )),
                 correlation_id: msg_id,
-                id: MessageId::in_response_to(&msg_id),
-            },
+                id: MessageId::in_response_to(&msg_id), // MessageId::new(), //
```
Comment (maqi) [inline src/node/interaction.rs:66]: forget to remove?

File: src/node/messaging.rs [hunk 5]
```diff
@@ -26,19 +31,105 @@ pub(crate) async fn send(msg: OutgoingMsg, network: &Network) -> Result<()> {
...
         |err| {
             error!("Unable to send msg: {:?}", err);
-            Err(Error::Logic(format!("Unable to send msg: {:?}", msg.id())))
+            Err(Error::UnableToSend(Msg::Client(message)))
+        },
+        |()| Ok(()),
+    )
+}
+
+// TODO: Refactor over support/error
```
Comment (maqi) [inline src/node/messaging.rs:100]: what does this TODO mean?
there is separate send_support and send_error

File: src/node/messaging.rs [hunk 6]
```diff
@@ -26,19 +31,105 @@ pub(crate) async fn send(msg: OutgoingMsg, network: &Network) -> Result<()> {
...
+    trace!("Sending support msg: {:?}", msg);
+    let src = SrcLocation::Node(network.our_name().await);
+    let itinerary = Itinerary {
+        src,
+        dst: msg.dst,
+        aggregation: Aggregation::None,
+    };
+
+    let dst_name = msg.dst.name().ok_or(Error::NoDestinationName)?;
+    let target_section_pk = network.get_section_pk_by_name(&dst_name).await?;
```
Comment (maqi) [inline src/node/messaging.rs:111]: is the variable name misleading or the function name?
why need the additional assignment from section_pk.bls() ?

File: src/node/messaging.rs [hunk 7]
```diff
@@ -26,19 +31,105 @@ pub(crate) async fn send(msg: OutgoingMsg, network: &Network) -> Result<()> {
...
+
+    let message = ClientMsg::SupportingInfo(msg.msg);
+    let result = network
+        .send_message(itinerary, message.serialize(dst_name, target_section_pk)?)
+        .await;
+
+    result.map_or_else(
+        |err| {
+            error!("Unable to send msg: {:?}", err);
+            Err(Error::UnableToSend(Msg::Client(message)))
```
Comment (maqi) [inline src/node/messaging.rs:125]: just to confirm, using Msg::Client is correct?
also, this mapping of result and the sending logic got duplicated for couple of times.
Is it possible to make it as a utility function like previous?

File: src/node/mod.rs [hunk 8]
```diff
@@ -147,45 +153,84 @@ impl Node {
...
+            // The message that triggered this error
+            MsgContext::Msg { msg, src } => {
+                warn!("Sending in response to a message: {:?}", msg);
+                let dst = get_dst_from_src(src);
+                match msg {
+                    Msg::Client(ClientMsg::Process(msg)) => {
+                        let error_message: ErrorMessage = match convert_to_error_message(err) {
+                            Ok(err) => err,
+                            Err(error) => {
+                                error!("Problem handling error: {:?}", error);
```
Comment (maqi) [inline src/node/mod.rs:194]: "When handling error {:?}, got problem {:?}", err, error

File: src/node/mod.rs [hunk 9]
```diff
@@ -147,45 +153,84 @@ impl Node {
...
+                            }
+                        };
+
+                        NodeDuty::SendError(OutgoingLazyError {
+                            msg: msg.create_processing_error(Some(error_message)),
+                            dst,
+                        })
+                    }
+                    Msg::Client(ClientMsg::ProcessingError(_err)) => {
+                        // TODO: handle error as a result of handling processing error...
```
Comment (maqi) [inline src/node/mod.rs:205]: just to confirm whether these TODOs must be resolved before merge?

File: src/node_ops.rs [hunk 10]
```diff
@@ -180,9 +185,15 @@ pub enum NodeDuty {
     SetNodeJoinsAllowed(bool),
     /// Send a message to the specified dst.
     Send(OutgoingMsg),
+    /// Send a lazy error as a result of a specific message.
+    /// The aim here is for the sender to respond with any missing state
+    SendError(OutgoingLazyError),
+    /// Send supporting in for for a given processing error.
```
Comment (maqi) [inline src/node_ops.rs:191]: Send supporting for a given processing error.

File: src/event_mapping/map_msg.rs [hunk 11]
```diff
@@ -128,56 +131,111 @@ pub fn match_user_sent_msg(msg: Message, origin: EndUser) -> Mapping {
...
+    }
+}
+
+/// Map a process error to relevant node duties
+pub fn map_node_process_err_msg(
+    msg: ProcessingError,
+    src: SrcLocation,
+    dst: DstLocation,
+) -> Mapping {
+    // debug!(" Handling received process err msg. {:?}.", msg);
```
Comment (maqi) [inline src/event_mapping/map_msg.rs:167]: remove this?

File: src/event_mapping/map_msg.rs [hunk 12]
```diff
@@ -128,56 +131,111 @@ pub fn match_user_sent_msg(msg: Message, origin: EndUser) -> Mapping {
...
+    match &dst {
+        DstLocation::Section(_name) | DstLocation::Node(_name) => match_process_err(msg, src),
```
Comment (maqi) [inline src/event_mapping/map_msg.rs:169]: DstLocation::Section(_)  | DstLocation::Node(_)

File: src/event_mapping/map_msg.rs [hunk 13]
```diff
@@ -128,56 +131,111 @@ pub fn match_user_sent_msg(msg: Message, origin: EndUser) -> Mapping {
...
+                msg: Msg::Client(ClientMsg::ProcessingError(msg)),
+                src,
+            },
+        },
+    }
+}
+
+fn match_process_err(msg: ProcessingError, src: SrcLocation) -> Mapping {
+    if let Some(reason) = msg.reason() {
+        // debug!("ProcessingError with reason")
```
Comment (maqi) [inline src/event_mapping/map_msg.rs:182]: remove this commented out block?
same comment to below

File: src/event_mapping/map_msg.rs [hunk 14]
```diff
@@ -128,56 +131,111 @@ pub fn match_user_sent_msg(msg: Message, origin: EndUser) -> Mapping {
...
+            src,
+        },
     }
 }
 
-fn match_or_err(msg: Message, src: SrcLocation) -> Mapping {
-    match match_node_msg(msg.clone(), src) {
-        NodeDuty::NoOp => Mapping::Error(LazyError {
+fn match_or_err(msg: NodeMsg, src: SrcLocation) -> Mapping {
+    // let msg_id = msg.id();
```
Comment (maqi) [inline src/event_mapping/map_msg.rs:208]: remove

File: src/node/messaging.rs (hunk 7 above)
Comment (oetyng) [reply src/node/messaging.rs:125]: > just to confirm, using Msg::Client is correct?

Should be. The message is for a client (`let message = ClientMsg::SupportingInfo(msg.msg);`).

File: src/node_ops.rs (hunk 10 above)
Comment (oetyng) [reply src/node_ops.rs:191]: Looks like that is supposed to be "Send supporting _info_ for".

File: src/node/messaging.rs (hunk 5 above)
Comment (oetyng) [reply src/node/messaging.rs:100]: IDK actually what Josh had in mind there.

File: src/node/mod.rs (hunk 9 above)
Comment (oetyng) [reply src/node/mod.rs:204]: It looks like these parts are necessary for the overall feature yes. I will need to dive into what the intention was here, to get it in place.

File: src/node/messaging.rs [hunk 15]
```diff
@@ -26,19 +31,105 @@ pub(crate) async fn send(msg: OutgoingMsg, network: &Network) -> Result<()> {
...
+    let itinerary = Itinerary {
+        src,
+        dst: msg.dst,
+        aggregation: Aggregation::None,
+    };
+
+    let dst_name = msg.dst.name().ok_or(Error::NoDestinationName)?;
+    let target_section_pk = network.get_section_pk_by_name(&dst_name).await?;
+
+    let target_section_pk = target_section_pk
```
Comment (maqi) [inline src/node/messaging.rs:113]: better cascade within the above line.

Comment (maqi) [inline src/node/messaging.rs:113]: better cascade within the above line.

File: src/node_ops.rs [hunk 16]
```diff
@@ -185,9 +189,15 @@ pub enum NodeDuty {
     SetNodeJoinsAllowed(bool),
     /// Send a message to the specified dst.
     Send(OutgoingMsg),
+    /// Send a lazy error as a result of a specific message.
+    /// The aim here is for the sender to respond with any missing state
+    SendError(OutgoingLazyError),
+    /// Send supporting info for a given processing error.
+    /// This should be any missing state required to proceed at the erring node.
+    SendSupport(OutgoingSupportingInfo),
```
Comment (maqi) [inline src/node_ops.rs:197]: Is that `SendError` is always from Node to Client,
meanwhile `SendSupport` is from Client to Node ?

File: src/event_mapping/mod.rs [hunk 17]
```diff
@@ -46,19 +52,63 @@ pub async fn map_routing_event(event: RoutingEvent, network_api: &Network) -> Ma
...
+                Mapping::Ok {
+                    op: NodeDuty::NoOp,
+                    ctx: Some(MsgContext::Msg {
+                        msg: Msg::Client(ClientMsg::ProcessingError(error)),
+                        src: SrcLocation::EndUser(user),
+                    }),
+                }
+            }
+            ClientMsg::SupportingInfo(msg) => Mapping::Ok {
+                op: NodeDuty::NoOp,
```
[the remark this answers is not in the document]
Comment (oetyng) [reply src/event_mapping/mod.rs]: Correct :+1:

File: src/node_ops.rs (hunk 16 above)
Comment (oetyng) [reply src/node_ops.rs:197]: I think `SendSupport` is from Node to Node.

File: src/chunks/chunk_storage.rs [hunk 18]
```diff
@@ -85,26 +87,21 @@ impl ChunkStorage {
...
             .get_chunk(address)
             .map_err(|_| ErrorMessage::DataNotFound(DataAddress::Blob(*address)));
 
-        // Sent back to data's metadata section, who will then
-        // forward it to client after having recorded the adult liveness.
-        ops.push(NodeDuty::Send(OutgoingMsg {
-            msg: Message::QueryResponse {
+        NodeDuty::Send(OutgoingMsg {
+            msg: MsgType::Node(NodeMsg::NodeQueryResponse {
+                response: NodeQueryResponse::Data(NodeDataQueryResponse::GetChunk(result)),
```
[the remark this answers is not in the document]
Comment (oetyng) [reply src/chunks/chunk_storage.rs:97]: It's just a categorization of different response types.

File: src/node/messaging.rs (hunk 6 above)
Comment (oetyng) [reply src/node/messaging.rs:111]: > is the variable name misleading or the function name?

Not sure what you refer to here.

File: src/node/mod.rs (hunk 9 above)
Comment (oetyng) [reply src/node/mod.rs]: Yeah, as you already noted in later comments, these shouldn't even be handled at Nodes, they are exclusively for Clients.

**Review by dirvine — APPROVED** [review]

**Comment by dirvine** [comment]
Merging to allow everyone to work on this

**Pull request merged**